docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+## Toolchain
+
+`Vector128`, `Vector128<T>`, `Vector256` and `Vector256<T>` expose a LOT of APIs. We are constrained by time, so we won't describe all of them with examples. Instead, we have grouped them into categories to give you an overview of their capabilities. It's not required to remember what each of these methods is doing, but it's important to remember what kind of operations they allow for and check the details when needed.
+
+**Note:** all of these methods have "software fallbacks", which are executed when they cannot be vectorized on given platform.
+
+### Creation
+
+Each of the vector types provides a `Create` method that accepts a single value and returns a vector with all elements initialized to this value.
+
+```cs
+public static Vector128<T> Create<T>(T value) where T : struct;
+```
+
+`CreateScalar` initializes first element to the specified value, and the remaining elements to zero.
+
+```cs
+public static Vector128<int> CreateScalar(int value);
+```
+
+`CreateScalarUnsafe` is similar, but the remaining elements are left uninitialized. It's dangerous!
+
+
+We also have an overload that allows for specifying every value in given vector:
+
+```cs
+public static Vector128<short> Create(short e0, short e1, short e2, short e3, short e4, short e5, short e6, short e7)
+```
+
+And last but not least we have a `Create` overload which accepts a buffer. It creates a vector with its elements set to the first `VectorXYZ<T>.Count` elements of the buffer. It's not recommended to use it in a loop, where `Load` methods should be used instead (for performance).
+
+```cs
+public static Vector128<T> Create<T>(ReadOnlySpan<T> values) where T : struct
+```
+
+to perform a copy in the other direction, we can use one of the `CopyTo` extension methods:
+
+```cs
+public static void CopyTo<T>(this Vector128<T> vector, Span<T> destination) where T : struct
+```
+
+### Bit operations
+
+All size-specific vector types provide a set of APIs for common bit operations.
+
+`BitwiseAnd` computes the bitwise-and of two vectors, `BitwiseOr` computes the bitwise-or of two vectors. They can both be expressed by using the corresponding operators (`&` and `|`). The same goes for `Xor` which can be expressed with `^` operator and `Negate` (`~`).
+
+```cs
+public static Vector128<T> BitwiseAnd<T>(Vector128<T> left, Vector128<T> right) where T : struct => left & right;
+public static Vector128<T> BitwiseOr<T>(Vector128<T> left, Vector128<T> right) where T : struct => left | right;
+public static Vector128<T> Xor<T>(Vector128<T> left, Vector128<T> right) => left ^ right;
+public static Vector128<T> Negate<T>(Vector128<T> vector) => ~vector;
+```
+
+`AndNot` computes the bitwise-and of a given vector and the ones' complement of another vector.
+
+```cs
+public static Vector128<T> AndNot<T>(Vector128<T> left, Vector128<T> right) => left & ~right;
```

Comment:
The pattern `left & ~right` should also get implicitly optimized as `AndNot` in .NET 8+